MI-220 File caching for user profile picture

limphilip opened the pull request:
A file caching is required for files that can be long to convert from the response of the server, such as the facebook user profile picture. Two options of caching were available:
- **Memory Caching**: Allows for faster retrieval of cached items, but less storage capacity.
- **Disk Storage Caching**: Allows to cache larger amount of items, but has slower performance compared to memory caching.

Memory caching was chosen, because the main goal here is to boost the performance of the caching. Although the storage capacity is limited (to 4MB in this case, but can easily be updated), the capacity can still hold about 26 user different profile pictures, which seems reasonable for the mRides application.

**Here is a summary of how the caching works**:
The caching mimics the Android LRU cache. The cache is represented by a dictionary in which the key is the facebook ID of the user associated with the picture being cached, and the value is a bitmap object representing the user profile picture. 

Due to the limited size of the cache, we need to decide which element is to be removed from the cache whenever it is full, and a new element needs to be added. The cache will remove the element that has not been accessed for the longest period of time. This is handled by a list of facebook user ID stored in which the beginning of the list represents the elements that were most recently accessed, and the end contains the elements that were accessed a long time ago.

**Integrating the cache with existing components**:
The `UserMapper` is in charge of retrieving the profile picture, either from the server or from the cache. Thus, the `UserMapper` first looks in the cache if the picture of a user is present, if it is, it will use it. Only when it is absent from the cache, that the mapper will ask the gateway to obtain it from the web.

The `UserMapper` asks the `UserCache` for the picture. Since the `UserCache` is already responsible for retrieving other types of cached information such the preferences of the user, it will have visibility over the new `MemoryCache` package, which is the actual implementation of the mimicked LRU cache.

**Testing**:
Unit tests and integration tests have been written in order to ensure that the cache functions properly. Those tests tests that functionalities of adding, removing and retrieving elements from the cache. They also ensure that the list of most recently accessed works properly, and that correct elements are removed from the cache when the cache is full and a new element needs to be added.

Also related to #47

limphilip committed:
MI-214 Initial implementation of cache for user profile picture

Since it is relatively fast already to retrieve, convert and display a
user profile picture, we do not need to save a lot of a data for this
task. Disk storage would not benefit by a lot in comparison with calling
the facebook api to obtain the picture. Rather, a memory cache limited
to 4MB (~26 pictures) is used. When the memory cache is full, the last
accessed picture is remove from the cache and replaced with a new
picture on the top of the list. This avoids calling the server
unnecessarily, keeping performance maximal and avoiding fetching data to
the disk, which would not be that advantageous compared to obtaining
data from the web.

The Cache has been added under a MemoryCache package.

The UserCache has visibility to this new Memory Cache.

The UserMapper consults the new interface of the user cache in order to
see whether the profile picture is in the cache before calling the
gateways.

limphilip committed:
MI-214 Path updated in csproj for MemoryCache package

Path updated in csproj for Cache/MemoryCache package. It is now in
MemoryCache and not directly in Cache directory.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-220/FileCaching

limphilip committed:
MI-214 MI-215 Added unit tests and integration tests for the UserFacebookProfilePictureCache and corrected bugs found by tests

- UserFacebookProfilePictureCache has been updated with the following:
the constant is now public for the max size of the cache, whenever
removing an element, the current size of the cache is now decreased as
well.

- Unit tests have been added and integration tests as well. They test
adding, removing, and retrieving pictures from the cache. They also test
whether the cache handles the most recently accessed list properly.

AlineTheAlien commented on UnitTests/CacheTest/MemoryCacheTest/UserFacebookProfilePictureCacheTest.cs:
```diff
@@ -0,0 +1,304 @@
+            // Check what we can find from the cache
+            Bitmap bitmapFound = userFbPicCache.FindUserFacebookProfilePicture(facebookId);
+            Assert.True(bitmapFound != null);
+            Assert.True(bitmapFound.Equals(imageBitmap));
```
It's bad practice to have a test case with more than one "assert", you might want to combine these into one "assert" statement. It helps pinpoint why a test case failed, and it follows the good quality assurance standards 👍 
Besides this the test cases look really good!

limphilip replied:
Thanks @AlineTheAlien , I will update this.

limphilip committed:
MI-215 Updated test cases with more significant Asserts

Updated the test cases with assertion that are more combined in order to
make it easier to pinpoint the problem if any.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-220/FileCaching

limphilip commented:
@AlineTheAlien I have updated the test cases and also updated myself with dev. Let me know if you can find any other issues, otherwise I think it's good to merge. Thanks again for pointing out the test cases 😃

AlineTheAlien commented:
@limphilip There seems to be a merge conflict in the app config file preventing me from merging, should I undo your last push (the merge of dev into your branch)? I think this branch was already safe to merge because there were no conflicts with dev so I think it's fine to merge this pull request without merging dev into it first.

limphilip commented:
It seems like my local dev wasn't up to date with the one on the server. I will update this branch again, and push when it's ready.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-220/FileCaching

# Conflicts:
#	mRides-app/mRides-app.csproj

limphilip committed:
Missing reference from the unit test project to the app

For some reasons the reference was broken/not working. I just removed
it, and re-added it again from unit test to mrides app project.

limphilip commented:
@AlineTheAlien It is up to date now, the conflict was due to files being added to the csproj, so they got added to the same lines. I took both the one from dev and mine so that all files are present when loading the visual studio project.

AlineTheAlien commented:
Ok looks good! Merging now 👍
